### DISCUSSION TITLE
How to start an extension project in visual studio to capture Debug.Start and Debug.StartWithoutDebugging events

### DISCUSSION THREAD

**Discussion by Ncerna** [VsixCommunity/Community.VisualStudio.Toolkit#411] [Q&A] [answered]
write visual studio extension how to capture Debug.Start and Debug.StartWithoutDebugging events..someone help me create a simple project to get started thanks

**Comment by RobertvanderHulst** [comment, marked as answer]
What I did is this:
- create a class that implements IVsDebuggerEvents
- get the debugger object with the help from the VS class inside the toolkit (must be done on the main thread)
```
await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
m_debugger = await VS.GetServiceAsync<SVsShellDebugger,IVsDebugger>();
Microsoft.VisualStudio.ErrorHandler.ThrowOnFailure(hr);
```
- register the class as event handler
 hr = m_debugger.AdviseDebuggerEvents(this, out m_Debuggercookie);

- get the initial value of the debugger 'state'
```
DBGMODE[] modeArray = new DBGMODE[1];
hr = m_debugger.GetMode(modeArray);
```
- The event handler looks like this
 ```
public int OnModeChange(DBGMODE dbgmodeNew)
{
    XSettings.DebuggerMode = (DebuggerMode)dbgmodeNew;
    return VSConstants.S_OK;
}
```
DebuggerMode is an enum in the toolkit with the values:
```
public enum DebuggerMode
    {
        Design = 0,
        Break = 1,
        Running = 2,
        EditAndContinue = 0x10000000,
        EditAndContinueMask = -268435457
    }
```
Don't forget to unregister the event handler when you no longer need it

```
await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
if (m_debugger != null && m_Debuggercookie != 0)
{
    hr = m_debugger.UnadviseDebuggerEvents(m_Debuggercookie);
    Microsoft.VisualStudio.ErrorHandler.ThrowOnFailure(hr);
}
```

**Reply by Ncerna** [reply]
to implement class IVsDebuggerEvents which kind of project I create from QUI..

![image](https://user-images.githubusercontent.com/54686131/215926409-14b14697-4171-4366-a5c5-89b5c5aca153.png)

**Reply by RobertvanderHulst** [reply]
You create a package. Inside the package, you register the event handler.
Or you download the source code for the Community Toolkit. There is a sample project in there.
You could try to add the debugger monitoring to the example

**Reply by Ncerna** [reply]
a favor and how do I initialize the package or the event to start capturing.

**Reply by Ncerna** [reply]
![image](https://user-images.githubusercontent.com/54686131/216777464-bf9997f9-f5a1-4bdd-99c2-c8adadfa9d3a.png)
 class that implements IVsDebuggerEvents , please give me a hand to be able to initialize

https://github.com/Ncerna/TestConcept/blob/main/Toolkit/Debugger/DebuggerEvents.cs

**Comment by RobertvanderHulst** [comment]
I would recommend creating an object of this class from within the InitializeAsync() that you already have in your package.
Did you also add the OnModeChange() event method.
And I would not use a discard variable (_) for the cookie. You are supposed to store that cookie, so you can use that cookie to unregister the event handler when needed.

**Reply by Ncerna** [reply]
I got it, thank you very much for your help...🤝

**Comment by Ncerna** [comment]
I want to write an extension for visual studio that captures the metrics (time, size and defect) of the Personal Software Process (PSP) as I would start... thanks a lot in advance ..
metrics:
Base LOC from a previous version.
Deleted .
Modified .
Added New .
Reused LOC.
New & Changed .
Total LOC The total program LOC
Total New Reused
🙏🙏🙏
